src/composites.rs

```diff
@@ -0,0 +1,214 @@
+use hyper::{Request, Response, StatusCode};
+use futures::{future, Future};
+use context::Context;
+
+/// Trait for getting the path of a request
+pub trait GetPath {
+    /// Retrieve the path
+    fn path(&self) -> &str;
+}
+
+impl GetPath for Request {
+    fn path(&self) -> &str {
+        self.path()
+    }
+}
+
+impl GetPath for (Request, Context) {
+    fn path(&self) -> &str {
+        self.0.path()
+    }
+}
+
+/// Trait for generating a default "not found" response
+pub trait NotFound {
+    /// Return a "not found" response
+    fn not_found() -> Self;
+}
+
+impl NotFound for Response {
+    fn not_found() -> Self {
+        Response::new().with_status(StatusCode::NotFound)
+    }
+}
+
+/// Trait for wrapping hyper NewServices to make the return type of new_service() uniform.
+/// This is necessary in order for the NewServices with different Instance types to
+/// be stored in a single collection.
+pub trait BoxedNewService<U, V, W> {
+    /// Create a new Service trait object
+    fn boxed_new_service(
+        &self,
+    ) -> Result<
+        Box<
+            Service<
+                Request = U,
+                Response = V,
+                Error = W,
+                Future = Box<Future<Item = V, Error = W>>,
+            >,
+        >,
+        io::Error,
+    >;
+}
+
+impl<T, U, V, W> BoxedNewService<U, V, W> for T
+where
+    T: NewService<Request = U, Response = V, Error = W>,
+    T::Instance: Service<Future = Box<Future<Item = V, Error = W>>>
+        + 'static,
+{
+    /// Call the new_service() method of the wrapped NewService and Box the result
+    fn boxed_new_service(
+        &self,
+    ) -> Result<
+        Box<
+            Service<
+                Request = U,
+                Response = V,
+                Error = W,
+                Future = Box<Future<Item = V, Error = W>>,
+            >,
+        >,
+        io::Error,
+    > {
+        let service = self.new_service()?;
+        Ok(Box::new(service))
+    }
+}
+
+/// A struct combining multiple NewServices. The NewServices are stored in a list
+/// in the order they are added, together with an associated base path. The service
+/// generated by the CompositeNewService will pass a request to the first service
+/// in the list whose base path is a prefix of the request's path or return a
+/// "not found" response if there is no match.
+pub struct CompositeNewService<U, V, W>(Vec<(&'static str, Box<BoxedNewService<U, V, W>>)>)
+where
+    U: GetPath,
+    V: NotFound + 'static,
+    W: 'static;
+
+impl<U: GetPath, V: NotFound, W> CompositeNewService<U, V, W> {
+    /// Create an empty CompositeNewService
+    pub fn new() -> Self {
+        CompositeNewService(Vec::new())
+    }
+
+    /// Add a new NewService with a base path to the composite
+    pub fn append_new_service(
+        &mut self,
+        base_path: &'static str,
+        new_service: Box<BoxedNewService<U, V, W>>,
+    ) {
+        self.0.push((base_path, new_service));
+    }
```

Comment:
Might be worth looking into having CompositeNewService implement `Deref<Target = Vec<(&'static str, Box<BoxedNewService<U, V, W>>)>>` https://doc.rust-lang.org/std/ops/trait.Deref.html#more-on-deref-coercion

That way people can just call straight to `my_composite_new_service.push((base_path, new_service));`, and in addition get all the other methods on `Vec` without needing to explicitly add them here when needed.  Not necessary, but worth thinking about.

Reply:
👍

Reply:
Now that you've added this, you can entirely remove `CompositeNewService#append_new_service` right?

Reply:
Needed to implement `DerefMut` as well, but after that, yeah.